sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -119,4 +121,202 @@ object IntervalUtils {
+      } else {
+        (m.group(9) + "000000000").substring(0, 9)
+      }
+      var nanos = toLongWithRange("nanosecond", nanoStr, 0L, 999999999L)
+      to match {
+        case "hour" =>
+          minutes = 0
+          seconds = 0
+          nanos = 0
+        case "minute" =>
+          seconds = 0
+          nanos = 0
+        case "second" =>
+          // No-op
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Cannot support (interval '$s' $from to $to) expression")
+      }
+      new CalendarInterval(0, sign * (
+        days * DateTimeUtils.MICROS_PER_DAY + hours * MICROS_PER_HOUR +
+        minutes * MICROS_PER_MINUTE + seconds * DateTimeUtils.MICROS_PER_SECOND +
+        nanos / 1000L))
+    } catch {
+      case e: Exception =>
+        throw new IllegalArgumentException(
+          "Error parsing interval day-time string: " + e.getMessage, e)
+    }
+  }
+
+  def fromUnitStrings(units: Array[String], values: Array[String]): CalendarInterval = {
+    assert(units.length == values.length)
+    var months: Int = 0
+    var microseconds: Long = 0
+    var i = 0
+    while (i < units.length) {
+      try {
+        units(i) match {
+          case "year" =>
+            months = Math.addExact(months, Math.multiplyExact(values(i).toInt, 12))
+          case "month" =>
+            months = Math.addExact(months, values(i).toInt)
+          case "week" =>
+            val weeksUs = Math.multiplyExact(values(i).toLong, 7 * DateTimeUtils.MICROS_PER_DAY)
+            microseconds = Math.addExact(microseconds, weeksUs)
+          case "day" =>
+            val daysUs = Math.multiplyExact(values(i).toLong, DateTimeUtils.MICROS_PER_DAY)
+            microseconds = Math.addExact(microseconds, daysUs)
+          case "hour" =>
+            val hoursUs = Math.multiplyExact(values(i).toLong, MICROS_PER_HOUR)
+            microseconds = Math.addExact(microseconds, hoursUs)
+          case "minute" =>
+            val minutesUs = Math.multiplyExact(values(i).toLong, MICROS_PER_MINUTE)
+            microseconds = Math.addExact(microseconds, minutesUs)
+          case "second" =>
+            microseconds = Math.addExact(microseconds, parseSecondNano(values(i)))
+          case "millisecond" =>
+            val millisUs = Math.multiplyExact(values(i).toLong, DateTimeUtils.MICROS_PER_MILLIS)
+            microseconds = Math.addExact(microseconds, millisUs)
+          case "microsecond" =>
+            microseconds = Math.addExact(microseconds, values(i).toLong)
+        }
+      } catch {
+        case e: Exception =>
+          throw new IllegalArgumentException("Error parsing interval string: " + e.getMessage, e)
+      }
+      i += 1
+    }
+    new CalendarInterval(months, microseconds)
+  }
+
+  /**
+   * Parse second_nano string in ss.nnnnnnnnn format to microseconds
+   */
+  private def parseSecondNano(secondNano: String): Long = {
+    val parts = secondNano.split("\\.")
+    if (parts.length == 1) {
```

Comment:
use a `match` statement maybe?

Reply:
I would do this refactoring in a separate PR too if you don't mind.